File: Heimdall/src/main/java/de/rheinfabrik/heimdall/grants/OAuth2Grant.java
package de.rheinfabrik.heimdall.grants;

import de.rheinfabrik.heimdall.OAuth2AccessToken;
import rx.Observable;

/**
 * Interface describing an OAuth2 Grant as described in https://tools.ietf.org/html/rfc6749#page-23.
 *
 * @param <TAccessToken> The access token type.
 */
public interface OAuth2Grant<TAccessToken extends OAuth2AccessToken> {

    // Abstract Api

    /**
     * Performs the actual request to grant a new access token.
     *
     * @return - An Observable emitting the granted access token.
     */
    Observable<TAccessToken> grantNewAccessToken();
}

File: library/src/main/java/de/rheinfabrik/heimdall2/grants/OAuth2Grant.kt
package de.rheinfabrik.heimdall2.grants

import de.rheinfabrik.heimdall2.OAuth2AccessToken
import io.reactivex.Single

/**
 * Interface describing an OAuth2 Grant as described in https://tools.ietf.org/html/rfc6749#page-23.
 *
 * @param <TAccessToken> The access token type.
 */
interface OAuth2Grant {

    // Abstract Api

    /**
     * Performs the actual request to grant a new access token.
     *
     * @return - An Observable emitting the granted access token.
     */
    fun grantNewAccessToken(): Single<OAuth2AccessToken>
}
